File: src/ustplayer/core/video_exporter.py
# video_exporter.py — 视频导出（封装 uPlRender 渲染器）
"""把 UST 工程渲染为 MP4 视频，并同步写入对应的 .uprd 工程文件。

流程：解析 UST → 由当前设置组装渲染配置（RenderConfig JSON）→ 写入 .uprd →
加载 `ustplayer_renderer.dll` 逐帧渲染 → 编码 MP4；可选再把伴奏音乐混入视频。
"""

import dataclasses
import json
import math
import os
import shutil
import subprocess
import tempfile
import time
from typing import Any, Callable, Optional

from ustplayer.core.contracts import APP_VERSION, NoteInfo, ProjectIO, UstInfo, UstParser
from ustplayer.core.log import logger
from ustplayer.core.renderer_ffi import RendererContext, RendererError, RendererLoader
from ustplayer.core.settings_manager import SettingsManager

# 伴奏混入失败时抛出的用户可读前缀，便于 UI 识别（ERcode011）
_AUDIO_MUX_FAILED = "音频混流失败"

# 子进程轮询间隔（秒）：取消请求最迟在此延迟后生效
_PROC_POLL_INTERVAL = 0.2


class VideoExporter:
    """实现 contracts.VideoExporter —— 通过 uPlRender DLL 渲染当前工程为视频。"""

    # 一拍 = 480 tick（与 ustPlayer / uPlRender 的约定一致）
    _TICKS_PER_QUARTER = 480

    def __init__(self, settings: SettingsManager, parser: UstParser, project_io: ProjectIO):
        self._settings = settings
        self._parser = parser
        self._project_io = project_io

    # ===================== 对外入口 =====================

    def render(
        self,
        output_path: str,
        width: int,
        height: int,
        fps: int,
        mux_audio: bool,
        progress_cb: Optional[Callable[[int], None]] = None,
        cancel_check: Optional[Callable[[], bool]] = None,
    ) -> str:
        """渲染当前工程为 MP4，并写入对应的 .uprd 工程文件。

        Returns:
            写入的 .uprd 工程文件路径。
        """
        output_path = self._ensure_mp4(output_path)
        core_ust = self._parse_ust()

        # 以「音频播完」为结束边界（无音频则按音符 tick 总长）：
        # 音符 tick 结束后、音频仍未播完的区间显示“空拍/静默文字”，音频播完后显示结束文字并停 1 秒。
        end_secs = self._render_end_secs(core_ust, cancel_check)
        render_ust = self._pad_trailing_rest(core_ust, end_secs)
        config = self._build_render_config(render_ust, output_path, width, height, fps)
        total_frames = self._frames_for(end_secs, fps)

        # 1) 写 .uprd → 2) 渲染 → 3) 可选混流；任一步失败/取消都清理半成品产物，
        # 避免残留打不开的 MP4 与指向无效视频的 .uprd
        uprd_path = self._uprd_path_for(output_path)
        try:
            self._project_io.export_uprd(uprd_path, {"width": width, "height": height, "fps": fps})
            logger.info(f"已写入 .uprd 工程: {uprd_path}")

            self._drive_renderer(config, render_ust, total_frames, progress_cb, cancel_check)

            if mux_audio and self._music_path:
                self._mux_audio(output_path, self._music_path, cancel_check)
        except BaseException:
            logger.warning(f"导出中止，清理半成品文件: {output_path}")
            self._remove_quiet(output_path)
            self._remove_quiet(output_path + ".mux.tmp.mp4")
            self._remove_quiet(uprd_path)
            raise

        return uprd_path

    # ===================== 数据准备 =====================

    def _ensure_mp4(self, output_path: str) -> str:
        path = output_path.strip()
        if not path.lower().endswith(".mp4"):
            path += ".mp4"
        return path

    def _parse_ust(self) -> UstInfo:
        ust_path = self._settings.file.ust_path.strip()
        if not ust_path or not os.path.exists(ust_path):
            raise FileNotFoundError(
                f"UST 文件不存在或未选择：{ust_path or '（空）'}"
            )
        return self._parser.parse(ust_path, self._settings.file.encoding)

    @property
    def _music_path(self) -> str:
        return self._settings.project.music_path.strip()

    def _lrc_text(self) -> Optional[str]:
        lrc_path = self._settings.player.lrc_path.strip()
        if not lrc_path or not os.path.exists(lrc_path):
            return None
        for enc in ("utf-8-sig", "utf-8", "gbk", "gb2312", "shift-jis"):
            try:
                with open(lrc_path, "r", encoding=enc) as f:
                    return f.read()
            except (UnicodeDecodeError, OSError):
                continue
        with open(lrc_path, "r", encoding="utf-8", errors="replace") as f:
            return f.read()

    def _build_render_config(
        self, core_ust: UstInfo, output_path: str, width: int, height: int, fps: int
    ) -> dict:
        """把当前设置 + UST 解析结果组装成渲染器需要的 RenderConfig JSON（dict）。"""
        params = self._settings.build_ust_info(core_ust)
        cfg = dataclasses.asdict(params)
        # 渲染器在 style.app_version 里画版权行；此处补充应用版本
        style = dict(cfg.get("style") or {})
        style["app_version"] = APP_VERSION
        cfg["style"] = style
        cfg["width"] = int(width)
        cfg["height"] = int(height)
        cfg["fps"] = int(fps)
        cfg["output_path"] = output_path
        return cfg

    @staticmethod
    def _uprd_path_for(output_path: str) -> str:
        stem, _ = os.path.splitext(output_path)
        return f"{stem}.uprd"

    @staticmethod
    def _remove_quiet(path: str) -> None:
        """尽力删除半成品文件；删除失败仅记日志，不掩盖导出流程的原异常。"""
        try:
            if os.path.exists(path):
                os.remove(path)
        except OSError as e:
            logger.warning(f"清理半成品文件失败: {path} ({e})")

    # ===================== 可取消的子进程执行 =====================

    @staticmethod
    def _run_cancellable(
        cmd: list,
        cancel_check: Optional[Callable[[], bool]],
        timeout_secs: float,
        stderr_file: Any = None,
    ) -> "tuple[int, str]":
        """Popen 启动子进程并轮询等待：期间持续响应 cancel_check 与超时。

        替代阻塞式 subprocess.run —— 否则进入 ffprobe/ffmpeg 阶段后，
        用户取消要等子进程自然结束（混流最长 1 小时）才能生效。
        stderr_file 可选传入以写模式打开的文件对象（如 tempfile.TemporaryFile），
        失败时可读取其内容辅助排查。
        返回 (退出码, stderr 尾部文本)；未捕获 stderr 时尾部为空字符串。
        """
        proc = subprocess.Popen(
            cmd,
            stdout=subprocess.DEVNULL,
            stderr=stderr_file or subprocess.DEVNULL,
        )
        deadline = time.monotonic() + timeout_secs
        while True:
            code = proc.poll()
            if code is not None:
                break
            if time.monotonic() >= deadline:
                proc.kill()
                proc.wait()
                raise TimeoutError(f"{os.path.basename(cmd[0])} 处理超时")
            if cancel_check and cancel_check():
                proc.kill()
                proc.wait()
                raise RuntimeError("导出已取消")
            time.sleep(_PROC_POLL_INTERVAL)
        tail = ""
        if stderr_file is not None:
            try:
                stderr_file.seek(0)
                tail = stderr_file.read()[-2000:].strip()
            except OSError:
                tail = ""
        return code, tail

    # ===================== 渲染驱动 =====================

    def _drive_renderer(
        self,
        config: dict,
        render_ust: UstInfo,
        total_frames: int,
        progress_cb: Optional[Callable[[int], None]],
        cancel_check: Optional[Callable[[], bool]],
    ):
        loader = RendererLoader(self._settings.program_root)
        lib = loader.load()
        config_json = json.dumps(config, ensure_ascii=False)
        ust_json = json.dumps(dataclasses.asdict(render_ust), ensure_ascii=False)

        with RendererContext(lib) as ctx:
            ctx.set_config(config_json)
            ctx.set_ust_text(ust_json)
            lrc_text = self._lrc_text()
            if lrc_text:
                ctx.set_lrc_text(lrc_text)

            ctx.begin_export()
            logger.info(f"开始渲染视频: {config['output_path']} (fps={config['fps']}, 总帧={total_frames})")
            for i in range(total_frames):
                if cancel_check and cancel_check():
                    raise RuntimeError("导出已取消")
                elapsed = i / config["fps"]
                ctx.render_frame(elapsed)
                if progress_cb and (i % 5 == 0 or i == total_frames - 1):
                    progress_cb(int((i + 1) / total_frames * 1000) if total_frames else 1000)
            ctx.end_export()

        logger.info("视频渲染完成（无声 MP4）")

    # ===================== 音频驱动的结束边界 =====================

    def _render_end_secs(self, core_ust: UstInfo, cancel_check: Optional[Callable[[], bool]] = None) -> float:
        """渲染结束边界（秒）：有伴奏时以“音频播完”为准（且不短于音符内容总长），
        无伴奏时按音符 tick 内容总长。"""
        content_secs = VideoExporter._content_secs(core_ust)
        audio_secs = self._audio_duration_secs(cancel_check)
        if audio_secs > 0:
            return max(content_secs, audio_secs)
        return content_secs

    @staticmethod
    def _content_secs(core_ust: UstInfo) -> float:
        """音符 tick 内容的时长（秒），与 timing.rs 一致：每音符长度下限 1 tick。"""
        tick_per_second = core_ust.tempo * VideoExporter._TICKS_PER_QUARTER / 60.0
        if tick_per_second <= 0:
            return 0.0
        total_tick = sum(max(max(n.length, 0), 1) for n in core_ust.notes)
        return total_tick / tick_per_second

    @staticmethod
    def _pad_trailing_rest(core_ust: UstInfo, end_secs: float) -> UstInfo:
        """在音符末尾补一个尾部休止音符（lyric=R），覆盖 [内容结束, 音频结束] 的空拍区间，
        使渲染器在该区间显示“空拍/静默文字”，并在音频结束时进入结束文字。"""
        content_secs = VideoExporter._content_secs(core_ust)
        if end_secs <= content_secs:
            return core_ust
        tick_per_second = core_ust.tempo * VideoExporter._TICKS_PER_QUARTER / 60.0
        if tick_per_second <= 0:
            return core_ust
        trailing_ticks = int(round((end_secs - content_secs) * tick_per_second))
        if trailing_ticks <= 0:
            return core_ust
        rest = NoteInfo(index="EXPORT_PAD", length=trailing_ticks, lyric="R", note_num=60)
        return UstInfo(
            version=core_ust.version,
            tempo=core_ust.tempo,
            tracks=core_ust.tracks,
            notes=[*core_ust.notes, rest],
        )

    @staticmethod
    def _frames_for(end_secs: float, fps: int) -> int:
        """总帧 = 结束边界(秒)*fps + 1 秒结束画面。"""
        base = int(math.ceil(end_secs * fps))
        return max(base + fps, 1)

    def _audio_duration_secs(self, cancel_check: Optional[Callable[[], bool]] = None) -> float:
        """用 ffprobe 读取伴奏音频时长（秒）；无伴奏 / 无 ffprobe / 失败时返回 0。"""
        mus = self._music_path
        if not mus or not os.path.exists(mus):
            return 0.0
        exe = shutil.which("ffprobe")
        if exe is None:
            return 0.0
        tmp = f"{mus}.dur.tmp"
        cmd = [
            exe,
            "-v", "error",
            "-show_entries", "format=duration",
            "-of", "default=noprint_wrappers=1:nokey=1",
            "-o", tmp,
            mus,
        ]
        try:
            code, _tail = VideoExporter._run_cancellable(cmd, cancel_check, 60)
            if code != 0:
                logger.warning(f"ffprobe 读取音频时长失败（退出码 {code}）")
                return 0.0
            with open(tmp, "r", encoding="utf-8") as f:
                text = f.read().strip()
            return float(text) if text else 0.0
        except RuntimeError:
            raise  # 用户取消必须向上传播，不能被吞成“时长未知”
        except Exception as e:
            logger.warning(f"ffprobe 读取音频时长失败: {e}")
            return 0.0
        finally:
            if os.path.exists(tmp):
                try:
                    os.remove(tmp)
                except OSError:
                    pass

    # ===================== 伴奏混流 =====================

    def _mux_audio(self, video_path: str, audio_path: str, cancel_check: Optional[Callable[[], bool]] = None):
        """用 ffmpeg 把伴奏音频混入无声 MP4（视频轨道复制，音频转 AAC）。

        stderr 落临时文件，失败时把尾部内容附进错误信息便于排查。"""
        if not os.path.exists(audio_path):
            logger.warning(f"伴奏文件不存在，跳过混流: {audio_path}")
            return
        exe = shutil.which("ffmpeg")
        if exe is None:
            raise RuntimeError(f"{_AUDIO_MUX_FAILED}：未找到 ffmpeg")
        tmp_path = f"{video_path}.mux.tmp.mp4"
        cmd = [
            exe, "-y",
            "-i", video_path,
            "-i", audio_path,
            "-c:v", "copy",
            "-c:a", "aac",
            "-map", "0:v:0",
            "-map", "1:a:0",
            "-movflags", "+faststart",
            tmp_path,
        ]
        with tempfile.TemporaryFile(mode="w+", encoding="utf-8", errors="replace") as errf:
            try:
                code, tail = VideoExporter._run_cancellable(
                    cmd, cancel_check, 3600, stderr_file=errf
                )
            except FileNotFoundError as e:
                raise RuntimeError(f"{_AUDIO_MUX_FAILED}：无法启动 ffmpeg") from e
            if code != 0:
                detail = f"：{tail}" if tail else ""
                raise RuntimeError(f"{_AUDIO_MUX_FAILED}：ffmpeg 退出码 {code}{detail}")
        os.replace(tmp_path, video_path)
        logger.info(f"已混入伴奏音频: {video_path}")


# 供 UI / 类型检测引用错误码（与 ERcode.txt 同步登记）
def renderer_error_message(exc: Exception) -> str:
    """把渲染相关异常转成面向用户的中文提示。"""
    if isinstance(exc, RendererError):
        return f"渲染器错误（{exc.code}）：{exc}"
    if isinstance(exc, FileNotFoundError):
        return str(exc)
    if isinstance(exc, RuntimeError):
        return str(exc)
    return str(exc)

File: ustPlayer/Video/VideoExporter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

using UstPlayer.Diagnostics;
using UstPlayer.Interop;
using UstPlayer.Models;
using UstPlayer.Projects;
using UstPlayer.Settings;
using UstPlayer.Ust;

namespace UstPlayer.Video;

/// <summary>
/// 视频导出 — 从 1.1.x <c>core/video_exporter.py</c> 的 <c>VideoExporter</c> 移植。
/// </summary>
/// <remarks>
/// <para>
/// 流程：解析 UST → 由当前设置组装渲染配置 → 写 <c>.uprd</c> → 调用渲染器逐帧渲染出无声 MP4
/// →（可选）用 ffmpeg 混入伴奏。任一步失败或取消都清理半成品，不留打不开的 MP4
/// 与指向无效视频的 <c>.uprd</c>。
/// </para>
/// <para>
/// <b>时序与播放器一致</b>（这是 1.1.x 重点对齐过的行为）：以「音频播完」为结束边界，
/// 音符内容结束而音频未播完的区间靠**补一个尾部休止音符</b>让渲染器显示空拍文字，
/// 音频播完后进入结束文字并保留 1 秒。
/// </para>
/// </remarks>
internal sealed class VideoExporter
{
    /// <summary>一拍（四分音符）的 tick 数，与播放器 / 渲染器一致。</summary>
    private const int TicksPerQuarterNote = 480;

    /// <summary>尾部休止音符的段编号（便于在渲染器日志里辨认）。</summary>
    private const string PaddingNoteIndex = "EXPORT_PAD";

    /// <summary>时长探测阶段的超时。</summary>
    private static readonly TimeSpan ProbeTimeout = TimeSpan.FromSeconds(60);

    private readonly SettingsManager _settings;
    private readonly UstFileReader _parser;
    private readonly UplrProjectIO _projectIo;
    private readonly IVideoMuxer _muxer;

    /// <summary>创建导出器。</summary>
    /// <param name="settings">设置管理器。</param>
    /// <param name="parser">UST 解析器。</param>
    /// <param name="projectIo">工程 IO（用于写 .uprd）。</param>
    /// <param name="muxer">音频混流器；传 <see langword="null"/> 用 ffmpeg 实现。</param>
    internal VideoExporter(
        SettingsManager settings,
        UstFileReader parser,
        UplrProjectIO projectIo,
        IVideoMuxer? muxer = null)
    {
        ArgumentNullException.ThrowIfNull(settings);
        ArgumentNullException.ThrowIfNull(parser);
        ArgumentNullException.ThrowIfNull(projectIo);

        _settings = settings;
        _parser = parser;
        _projectIo = projectIo;
        _muxer = muxer ?? new FfmpegMuxer(settings.ProgramRoot);
    }

    /// <summary>
    /// 把当前工程渲染为 MP4，并写入对应的 <c>.uprd</c> 工程文件。
    /// </summary>
    /// <param name="outputPath">输出 MP4 路径。</param>
    /// <param name="width">画面宽。</param>
    /// <param name="height">画面高。</param>
    /// <param name="fps">帧率。</param>
    /// <param name="muxAudio">是否把伴奏混入视频。</param>
    /// <param name="progressCallback">进度回调（千分比 0..1000）。</param>
    /// <param name="cancelCheck">取消检查；返回 <see langword="true"/> 时提前终止。</param>
    /// <param name="cancellationToken">取消令牌。</param>
    /// <returns>写入的 <c>.uprd</c> 路径。</returns>
    /// <exception cref="FileNotFoundException">UST 文件缺失。</exception>
    /// <exception cref="RendererException">渲染器缺失或渲染失败。</exception>
    /// <exception cref="OperationCanceledException">被取消。</exception>
    internal async Task<string> RenderAsync(
        string outputPath,
        int width,
        int height,
        int fps,
        bool muxAudio,
        Action<int>? progressCallback = null,
        Func<bool>? cancelCheck = null,
        CancellationToken cancellationToken = default)
    {
        var mp4Path = EnsureMp4Extension(outputPath);
        ValidateVideoParameters(width, height, fps);

        var coreUst = ParseUst();

        // 以「音频播完」为结束边界（无音频则按音符内容总长）
        var endSeconds = await ResolveEndSecondsAsync(cancelCheck, cancellationToken).ConfigureAwait(false);
        var renderUst = PadTrailingRest(coreUst, endSeconds);
        var totalFrames = FramesFor(endSeconds, fps);

        var uprdPath = UprdPathFor(mp4Path);

        try
        {
            _projectIo.ExportUprd(uprdPath, width, height, fps);
            AppLogger.Info($"已写入 .uprd 工程：{uprdPath}");

            DriveRenderer(renderUst, mp4Path, width, height, fps, totalFrames,
                progressCallback, cancelCheck, cancellationToken);

            if (muxAudio && !string.IsNullOrWhiteSpace(MusicPath))
            {
                await _muxer.MuxAsync(mp4Path, MusicPath, cancelCheck, cancellationToken)
                    .ConfigureAwait(false);
            }
        }
        catch (Exception)
        {
            // 任一步失败 / 取消都清理半成品，避免残留打不开的 MP4 与指向无效视频的 .uprd
            AppLogger.Warning($"导出中止，清理半成品文件：{mp4Path}");
            TryDelete(mp4Path);
            TryDelete(mp4Path + ".mux.tmp.mp4");
            TryDelete(uprdPath);
            throw;
        }

        return uprdPath;
    }

    /// <summary>混流器是否可用（UI 可据此提示用户）。</summary>
    internal bool CanMuxAudio => _muxer.IsAvailable;

    /// <summary>伴奏文件路径。</summary>
    private string MusicPath => _settings.Project.MusicPath.Trim();

    // ===================== 数据准备 =====================

    private UstInfo ParseUst()
    {
        var ustPath = _settings.File.UstPath.Trim();

        if (string.IsNullOrEmpty(ustPath) || !File.Exists(ustPath))
        {
            throw new FileNotFoundException($"UST 文件不存在或未选择：{(ustPath.Length == 0 ? "（空）" : ustPath)}");
        }

        return _parser.Parse(ustPath, _settings.File.Encoding);
    }

    /// <summary>确保输出路径以 <c>.mp4</c> 结尾。</summary>
    /// <param name="outputPath">用户给的路径。</param>
    /// <returns>规范化后的路径。</returns>
    internal static string EnsureMp4Extension(string outputPath)
    {
        var path = (outputPath ?? string.Empty).Trim();
        return path.EndsWith(".mp4", StringComparison.OrdinalIgnoreCase) ? path : path + ".mp4";
    }

    /// <summary>视频参数必须为正整数（否则后续可能除零）。</summary>
    /// <param name="width">宽。</param>
    /// <param name="height">高。</param>
    /// <param name="fps">帧率。</param>
    internal static void ValidateVideoParameters(int width, int height, int fps)
    {
        if (width <= 0 || height <= 0 || fps <= 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(width),
                $"视频参数必须为正数：width={width}, height={height}, fps={fps}");
        }
    }

    /// <summary><c>.uprd</c> 路径：与视频同名的同目录文件。</summary>
    /// <param name="outputPath">视频路径。</param>
    /// <returns>uprd 路径。</returns>
    internal static string UprdPathFor(string outputPath) =>
        Path.ChangeExtension(outputPath, ".uprd");

    /// <summary>音符内容的时长（秒）：每音符长度下限为 1 tick（与渲染器一致）。</summary>
    /// <param name="ust">UST 数据。</param>
    /// <returns>秒数。</returns>
    internal static double ContentSeconds(UstInfo ust)
    {
        ArgumentNullException.ThrowIfNull(ust);

        var ticksPerSecond = ust.Tempo * TicksPerQuarterNote / 60.0;
        if (ticksPerSecond <= 0)
        {
            return 0.0;
        }

        long totalTicks = 0;
        foreach (var note in ust.Notes)
        {
            totalTicks += Math.Max(Math.Max(note.Length, 0), 1);
        }

        return totalTicks / ticksPerSecond;
    }

    /// <summary>
    /// 在音符末尾补一个休止音符，覆盖 [内容结束, 音频结束] 的空拍区间。
    /// </summary>
    /// <param name="ust">原始 UST。</param>
    /// <param name="endSeconds">结束边界（秒）。</param>
    /// <returns>补过尾部休止的 UST（无需补时返回原对象）。</returns>
    /// <remarks>
    /// 播放器在「内容结束但音频未播完」区间显示空拍文字；导出要复现同一画面，
    /// 就得让渲染器在这段时间里也有内容可显示——补一个 <c>lyric=R</c> 的尾部音符即可。
    /// </remarks>
    internal static UstInfo PadTrailingRest(UstInfo ust, double endSeconds)
    {
        ArgumentNullException.ThrowIfNull(ust);

        var contentSeconds = ContentSeconds(ust);
        if (endSeconds <= contentSeconds)
        {
            return ust;
        }

        var ticksPerSecond = ust.Tempo * TicksPerQuarterNote / 60.0;
        if (ticksPerSecond <= 0)
        {
            return ust;
        }

        var trailingTicks = (int)Math.Round(
            (endSeconds - contentSeconds) * ticksPerSecond, MidpointRounding.AwayFromZero);

        if (trailingTicks <= 0)
        {
            return ust;
        }

        var notes = new List<NoteInfo>(ust.Notes.Count + 1);
        notes.AddRange(ust.Notes);
        notes.Add(new NoteInfo
        {
            Index = PaddingNoteIndex,
            Length = trailingTicks,
            Lyric = "R",
            NoteNumber = 60,
        });

        return new UstInfo
        {
            Version = ust.Version,
            Tempo = ust.Tempo,
            Tracks = ust.Tracks,
            Notes = notes,
        };
    }

    /// <summary>总帧数 = 结束边界 × 帧率 + 1 秒结束画面。</summary>
    /// <param name="endSeconds">结束边界（秒）。</param>
    /// <param name="fps">帧率。</param>
    /// <returns>帧数。</returns>
    internal static int FramesFor(double endSeconds, int fps)
    {
        var baseFrames = (int)Math.Ceiling(endSeconds * fps);
        return Math.Max(baseFrames + fps, 1);
    }

    /// <summary>读取 LRC 文本（多编码回退）；不存在返回 <see langword="null"/>。</summary>
    /// <returns>LRC 文本或 <see langword="null"/>。</returns>
    private string? ReadLrcText()
    {
        var lrcPath = _settings.Player.LrcPath.Trim();

        if (string.IsNullOrEmpty(lrcPath) || !File.Exists(lrcPath))
        {
            return null;
        }

        return ReadTextWithEncodingFallback(lrcPath);
    }

    /// <summary>读取伴奏音频时长（秒）；无伴奏 / 无 ffprobe / 失败时返回 0。</summary>
    /// <param name="cancelCheck">取消检查。</param>
    /// <param name="cancellationToken">取消令牌。</param>
    /// <returns>时长（秒）。</returns>
    private async Task<double> AudioDurationSecondsAsync(
        Func<bool>? cancelCheck,
        CancellationToken cancellationToken)
    {
        var musicPath = MusicPath;
        if (string.IsNullOrEmpty(musicPath) || !File.Exists(musicPath))
        {
            return 0.0;
        }

        var ffprobe = ExternalToolLocator.Find("ffprobe", _settings.ProgramRoot);
        if (ffprobe is null)
        {
            return 0.0;
        }

        string[] arguments =
        [
            "-v", "error",
            "-show_entries", "format=duration",
            "-of", "default=noprint_wrappers=1:nokey=1",
            musicPath,
        ];

        try
        {
            var result = await CancellableProcess.RunAsync(
                ffprobe, arguments, cancelCheck, ProbeTimeout, cancellationToken).ConfigureAwait(false);

            if (result.ExitCode != 0)
            {
                AppLogger.Warning($"ffprobe 读取音频时长失败（退出码 {result.ExitCode}）");
                return 0.0;
            }

            var text = result.StandardOutput.Trim();
            return double.TryParse(text, NumberStyles.Float, CultureInfo.InvariantCulture, out var seconds)
                ? seconds
                : 0.0;
        }
        catch (OperationCanceledException)
        {
            // 用户取消必须向上传播，不能被吞成「时长未知」
            throw;
        }
        catch (TimeoutException)
        {
            AppLogger.Warning("ffprobe 读取音频时长超时");
            return 0.0;
        }
        catch (Exception exception)
        {
            AppLogger.Warning($"ffprobe 读取音频时长失败：{exception.Message}");
            return 0.0;
        }
    }

    /// <summary>渲染结束边界（秒）。</summary>
    /// <param name="cancelCheck">取消检查。</param>
    /// <param name="cancellationToken">取消令牌。</param>
    /// <returns>秒数。</returns>
    private async Task<double> ResolveEndSecondsAsync(
        Func<bool>? cancelCheck,
        CancellationToken cancellationToken)
    {
        var coreUst = ParseUst();
        var contentSeconds = ContentSeconds(coreUst);
        var audioSeconds = await AudioDurationSecondsAsync(cancelCheck, cancellationToken).ConfigureAwait(false);

        return audioSeconds > 0 ? Math.Max(contentSeconds, audioSeconds) : contentSeconds;
    }

    // ===================== 渲染驱动 =====================

    /// <summary>驱动渲染器逐帧渲染。</summary>
    private void DriveRenderer(
        UstInfo renderUst,
        string outputPath,
        int width,
        int height,
        int fps,
        int totalFrames,
        Action<int>? progressCallback,
        Func<bool>? cancelCheck,
        CancellationToken cancellationToken)
    {
        var parameters = _settings.BuildLaunchParams(renderUst);
        var config = RenderConfig.Build(parameters, width, height, fps, outputPath);
        var ustJson = RenderConfig.BuildUstJson(renderUst);

        using var renderer = UplRenderContext.Create();

        // 渲染器的编码器只从 PATH 查找 ffmpeg；内置版本位于 <程序目录>/ffmpeg，
        // 必须在 begin_export 前临时加入 PATH。
        using var pathScope = BundledFfmpegPathScope.Enter(_settings.ProgramRoot);

        renderer.SetConfig(config);
        renderer.SetUstText(ustJson);

        if (ReadLrcText() is { } lrcText)
        {
            renderer.SetLrcText(lrcText);
        }

        renderer.SetProgressCallback(progressCallback);

        renderer.BeginExport();
        AppLogger.Info($"开始渲染视频：{outputPath}（fps={fps}，总帧={totalFrames}）");

        for (var index = 0; index < totalFrames; index++)
        {
            cancellationToken.ThrowIfCancellationRequested();

            if (cancelCheck?.Invoke() ?? false)
            {
                throw new OperationCanceledException("导出已取消");
            }

            renderer.RenderFrame(index / (double)fps);
        }

        renderer.EndExport();
        AppLogger.Info("视频渲染完成（无声 MP4）");
    }

    // ===================== 辅助 =====================

    private static string ReadTextWithEncodingFallback(string path)
    {
        foreach (var encoding in EncodingsToTry())
        {
            try
            {
                return File.ReadAllText(path, encoding);
            }
            catch (DecoderFallbackException)
            {
                // 换下一个编码
            }
        }

        return File.ReadAllText(path, Encoding.UTF8);
    }

    private static List<Encoding> EncodingsToTry()
    {
        var list = new List<Encoding>
        {
            new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true),
        };

        foreach (var name in (string[])["gbk", "gb2312", "shift-jis"])
        {
            try
            {
                list.Add(Encoding.GetEncoding(
                    name, EncoderFallback.ExceptionFallback, DecoderFallback.ExceptionFallback));
            }
            catch (ArgumentException)
            {
                // 该代码页不可用：跳过
            }
        }

        return list;
    }

    private static void TryDelete(string path)
    {
        try
        {
            if (File.Exists(path))
            {
                File.Delete(path);
            }
        }
        catch (Exception exception)
        {
            AppLogger.Warning($"清理半成品文件失败：{path}（{exception.Message}）");
        }
    }
}
